### PR TITLE
Snap yaml

### PR DESCRIPTION
**Pull request by sergiusens**
The documentation will be in a separate PR/commit

The bug this fixes is LP: [#1532842](https://bugs.launchpad.net/click-reviewers-tools/+bug/1532842)

### CODE REVIEW

File: snapcraft/meta.py [hunk 1]
```diff
@@ -230,17 +190,13 @@ def _wrap_exe(relexepath, args=None):
     wrappath = exepath + '.wrapper'
     shebang = None
 
-    # TODO talk to original author if the exception to be captured here is
-    # FileNotFoundError, the original code was a general catch all
-    try:
+    with contextlib.suppress(FileNotFoundError):
```
Comment (kyrofa) [inline snapcraft/meta.py]: Would it be more clear if you simply checked for the presence of the file before removing it?

**Comment by kyrofa** [comment]
This looks good to me.

File: snapcraft/main.py [hunk 2]
```diff
@@ -89,6 +89,7 @@ def main():
     try:
         commands.load(args['COMMAND']).main(argv=args['ARGS'])
     except Exception as e:
+        raise e
```
Comment (elopio) [inline snapcraft/main.py]: huh, this is weird. With this the sys exit will never happen.

File: snapcraft/meta.py [hunk 3]
```diff
@@ -1,6 +1,6 @@
 # -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
 #
-# Copyright (C) 2015-2016 Canonical Ltd
+# Copyright (C) 2016 Canonical Ltd
```
Comment (elopio) [inline snapcraft/meta.py]: Why did you remove the 2015? It should have all the years that were touched.

File: snapcraft/meta.py [hunk 4]
```diff
@@ -31,73 +35,67 @@
...
 
 def create(config_data):
-    '''
-    Create  the meta directory and provision it with package.yaml and readme.md
-    in the snap dir using information from config_data and arches.
-    If provided arches, is a list of arches.
-
-    Returns meta_dir.
-    '''
+    """Creates snap.yaml and necessary package hooks.
```
Comment (elopio) [inline snapcraft/meta.py]: s/Creates/Create

File: snapcraft/meta.py [hunk 5]
```diff
@@ -108,91 +106,53 @@ def _setup_config_hook(hooks_dir, config):
...
 
-    if config_data.get('license-agreement', '') == 'explicit':
-        package_yaml['explicit-license-agreement'] = 'yes'
+    return snap_yaml
 
-    return package_yaml
 
+def _copy(meta_dir, relpath, new_relpath=None):
+    new_base = new_relpath or os.path.basename(relpath)
+    new_relpath = os.path.join(meta_dir, new_base)
```
Comment (elopio) [inline snapcraft/meta.py]: nit, do not overwrite the argument, use a different name for the var.

File: snapcraft/meta.py [hunk 6]
```diff
@@ -263,26 +219,10 @@ def _wrap_exe(relexepath, args=None):
...
-        stoppath = services[name].get('stop')
-        if stoppath:
-            stopparts = shlex.split(stoppath)
-            stopwrap = _wrap_exe(stopparts[0])
-            services[name]['stop'] = _replace_cmd(stopparts, stopwrap)
+def _wrap_apps(apps):
+    for app in apps:
+        for k in [k for k in ('command', 'stop-command') if k in apps[app]]:
+            execparts = shlex.split(apps[app][k])
+            apps[app][k] = _wrap_exe(execparts[0], args=execparts[1:])
```
Comment (elopio) [inline snapcraft/meta.py]: nice refactor!

File: snapcraft/tests/test_meta.py [hunk 7]
```diff
@@ -27,331 +26,159 @@
 )
 
 
-class ComposeTestCase(tests.TestCase):
+class Create(tests.TestCase):
```
Comment (elopio) [inline snapcraft/tests/test_meta.py]: CreateTestCase

File: snapcraft/tests/test_meta.py [hunk 8]
```diff
@@ -27,331 +26,159 @@
...
-            call().__enter__().write("'"),
-            call().__enter__().write('\n'),
-            call().__enter__().flush(),
-            call().__enter__().flush(),
-            call().__exit__(None, None, None),
-            call(os.path.join(self.meta_dir, 'readme.md'), 'w'),
-            call().__enter__(),
-            call().__enter__().write('my summary\nmy description\n'),
-            call().__exit__(None, None, None),
-        ]
```
Comment (elopio) [inline snapcraft/tests/test_meta.py]: Yes, I like the removal of this :)

**Comment by elopio** [comment]
We are missing examples of forking daemons. I'll report a bug for this.

File: snapcraft/meta.py (hunk 1 above)
Comment (sergiusens) [inline snapcraft/meta.py]: yeah, I'll change that

File: snapcraft/main.py (hunk 2 above)
Comment (sergiusens) [inline snapcraft/main.py]: yeah, escaped me.

File: snapcraft/meta.py (hunk 3 above)
Comment (sergiusens) [inline snapcraft/meta.py]: because, not sure how it worked, but I moved `meta.py` to `meta_legacy.py` and created a new `meta.py` which only has 2016 things in it.

File: snapcraft/meta.py (hunk 4 above)
Comment (sergiusens) [inline snapcraft/meta.py]: k

File: snapcraft/meta.py (hunk 5 above)
Comment (sergiusens) [inline snapcraft/meta.py]: yeah, this is ... ... ... :-)

File: snapcraft/meta.py (hunk 6 above)
Comment (sergiusens) [inline snapcraft/meta.py]: thanks :-)
Simplified duplication

File: snapcraft/tests/test_meta.py (hunk 7 above)
Comment (sergiusens) [inline snapcraft/tests/test_meta.py]: k

File: snapcraft/tests/test_meta.py (hunk 8 above)
Comment (sergiusens) [inline snapcraft/tests/test_meta.py]: I thought you would like the new testing mechanism :wink:

**Comment by sergiusens** [comment]
I've updated the review taking into account the comments from @kyrofa and @elopio 
I've also added the bug marker to the the commit message

**Comment by elopio** [comment]
Looks good to me too.

**Comment by elopio** [comment]
And bug reported about the forking: https://bugs.launchpad.net/snapcraft/+bug/1532853
